pyhms/core/initializers.py

```diff
@@ -0,0 +1,120 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+from pyhms.core.individual import Individual
+from pyhms.core.problem import Problem
+from pyhms.utils.samplers import sample_normal, sample_uniform
+from scipy.stats.qmc import LatinHypercube, Sobol
+
+
+class PopInitializer(ABC):
+    def __init__(self, bounds: np.ndarray | None = None):
+        self._bounds = bounds
+
+    @abstractmethod
+    def sample_pop(self, pop_size: int | None, problem: Problem) -> list[Individual]:
+        pass
+
+    def __call__(self, pop_size: int | None, problem: Problem):
+        return self.sample_pop(pop_size, problem)
+
+
+class SeededPopInitializer(PopInitializer):
+    @abstractmethod
+    def get_seed(self, problem: Problem) -> Individual:
+        pass
+
+
+class UniformGlobalInitializer(PopInitializer):
+    def __init__(self, bounds: np.ndarray):
+        super().__init__(bounds)
+        self.sampler = sample_uniform(bounds)
+
+    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
+        return [Individual(genome=genome, problem=problem) for genome in self.sampler(pop_size)]
+
+
+class LHSGlobalInitializer(PopInitializer):
+    def __init__(self, bounds: np.ndarray, random_seed: int = None):
+        super().__init__(bounds)
+        self.sampler = LatinHypercube(d=len(bounds), seed=random_seed)
+        self.lower_bounds = bounds[:, 0]
+        self.upper_bounds = bounds[:, 1]
+
+    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
+        sample = self.sampler.random(pop_size)
+        genomes = self.lower_bounds + sample * (self.upper_bounds - self.lower_bounds)
+        return [Individual(genome, problem=problem) for genome in genomes]
+
+
+class SobolGlobalInitializer(PopInitializer):
+    def __init__(self, bounds: np.ndarray, random_seed: int = None):
+        super().__init__(bounds)
+        self.sampler = Sobol(d=len(bounds), scramble=True, seed=random_seed)
+        self.lower_bounds = bounds[:, 0]
+        self.upper_bounds = bounds[:, 1]
+
+    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
+        sample = self.sampler.random(pop_size)
+        genomes = self.lower_bounds + sample * (self.upper_bounds - self.lower_bounds)
+        return [Individual(genome, problem=problem) for genome in genomes]
```

Comment:
maybe we can have a method in `PopInitializer`:
```
def sample_to_individuals(self, sample: np.ndarray, problem: Problem) -> list[Individual]:
        return [Individual(genome=genome, problem=problem) for genome in sample]
```
and then we have:
```
class SobolGlobalInitializer(PopInitializer):
     def __init__(self, bounds: np.ndarray, random_seed: int = None):
        super().__init__(bounds)
        self.sampler = Sobol(d=len(bounds), scramble=True, seed=random_seed)

    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
        sample = self.sampler.random(pop_size)
        return self.sample_to_individuals(linear_scale(sample, problem.bounds), problem)
```

Reply:
and maybe all samplers can work like this? to just initialize the sampler and then you have the common sample_pop method:
```
def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
        sample = self.sampler.random(pop_size)
        return self.sample_to_individuals(linear_scale(sample, problem.bounds), problem)
```